### DISCUSSION TITLE
Memory reservation for nlohmann json

### DISCUSSION THREAD

**Discussion by sreerajchundayil** [nlohmann/json#3366] [Q&A]
Background:
---
I am populating nlohmann json object in memory and later writing to a file.
The json structure contains a lot of nesting up to 8 levels.
The memory usage can grow upto 20GB.

Now
----
I don't know the inner working of the nlohmann json memory allocations.

Is there any memory reserve mechanism for the nlohmann json object like we do for vector?
I expect the memory is assigned on the fly and some reallocation might be happening when huge data is populated.

OR
 Any performance improvement approach for such a huge data in memory?

**Comment by nlohmann** [comment]
- Which compiler and operating system are you using?
- Did you compile in Release or Debug mode?

**Comment by sreerajchundayil** [comment]
-> compiler = g++ (GCC) 7.3.0
-> OS = CentOS Linux 7 (Core)
-> Compiled in debug mode.

**Reply by nlohmann** [reply]
Can you please try in Release mode?

**Reply by sreerajchundayil** [reply]
I am not facing any issue with the current approach. I was just asking whether we can reserve memory like a vector and improve performance. I don't have any benchmark to say currently the performance is bad.

**Reply by nlohmann** [reply]
No, there is currently no API for this.

**Reply by nlohmann** [reply]
Well - you could get a pointer to the underlying vector with https://json.nlohmann.me/api/basic_json/get_ptr/ and then call `reserve` there.

**Reply by falbrechtskirchinger** [reply]
The SAX interface (https://json.nlohmann.me/api/json_sax/) could possibly be extended to pass a reference to `start_array` and `start_object`. One could then call `get_ptr` and `reserve`. Just an idea if anyone wants to take a closer look.

**Reply by sreerajchundayil** [reply]
Thanks.
